contracts/halo-pool/src/contract.rs

```diff
@@ -294,68 +562,122 @@ pub fn execute_withdraw(
-    if current_staker_amount < amount {
+    if staker_info.amount < amount {
         return Err(ContractError::Std(StdError::generic_err(
             "InsufficientFunds: Withdraw amount exceeds staked amount",
         )));
     }
 
+    // Init reward amount
+    let mut reward_amount = Uint128::zero();
+    // Get current pool info in pool infos
+    let pool_info = POOL_INFOS.load(deps.storage)?.pool_infos[current_phase_index as usize].clone();
+    // Init new accrued token per share
+    let mut new_accrued_token_per_share;
+    // Init new last reward time
+    let mut new_last_reward_time;
+
+    let mut res = Response::new();
+
+    // get staker joined phases
+    let staker_joined_phase = staker_info.joined_phase;
+
+    // If staker has joined previous phases, loops all pool infos to get reward per second from current pool index to staker joined phases
+    if staker_joined_phase < current_phase_index {
+        for i in staker_joined_phase..current_phase_index {
+            // Get last reward time
+            let last_reward_time = pool_infos.last_reward_time[i as usize];
+            // Get accrued token per share
+            let accrued_token_per_share = pool_infos.accrued_token_per_share[i as usize];
+            // Get pool info from pool infos
+            let pool_info = POOL_INFOS.load(deps.storage)?.pool_infos[i as usize].clone();
+            // Get staked token balance from pool contract previous phases
+            let staked_token_supply = pool_info.total_staked_at_end_time;
+
+            // update pool
+            (new_accrued_token_per_share, new_last_reward_time) = update_pool(
+                pool_info.end_time,
+                pool_info.reward_per_second,
+                staked_token_supply,
+                accrued_token_per_share,
+                pool_info.end_time,
+                last_reward_time,
+            );
+
+            pool_infos.last_reward_time[i as usize] = new_last_reward_time;
+            pool_infos.accrued_token_per_share[i as usize] = new_accrued_token_per_share;
+
+            // Save pool infos
+            POOL_INFOS.save(deps.storage, &pool_infos)?;
+
+            reward_amount += calc_reward_amount(
+                staker_info.amount,
+                new_accrued_token_per_share,
+                staker_info.reward_debt[i as usize],
+            );
+            // Update staker info
+            staker_info.reward_debt[i as usize] = staker_info.amount * new_accrued_token_per_share;
+        }
+        // Increase length of user reward debt to current pool index
+        staker_info.reward_debt.push(Uint128::zero());
+    }
+    // Get last reward time
+    let last_reward_time = pool_infos.last_reward_time[current_phase_index as usize];
+    // Get accrued token per share
+    let accrued_token_per_share = pool_infos.accrued_token_per_share[current_phase_index as usize];
     // Get staked token balance from pool contract
     let staked_token_supply = query_token_balance(
```

Comment:
cái này có thể ghi vào thành 1 trường data của phase luôn ko nhỉ. Mỗi lần cần tính lại query thì hơi mất thời gian

Reply:
Cái này đúng là cần đưa vào PoolInfos ạ, vì nếu gọi query thì có thể sẽ có trường hợp ai đó gửi thẳng LP vào contract dẫn đến tính toán sai ạ.
Moved